firebase-firestore/src/main/java/com/google/firebase/firestore/util/Util.java

```diff
@@ -87,9 +90,45 @@ public static int compareIntegers(int i1, int i2) {
 
   /** Compare strings in UTF-8 encoded byte order */
   public static int compareUtf8Strings(String left, String right) {
-    ByteString leftBytes = ByteString.copyFromUtf8(left);
-    ByteString rightBytes = ByteString.copyFromUtf8(right);
-    return compareByteStrings(leftBytes, rightBytes);
+    int i = 0;
+    while (i < left.length() && i < right.length()) {
+      int leftCodePoint = left.codePointAt(i);
+      int rightCodePoint = right.codePointAt(i);
+
+      if (leftCodePoint != rightCodePoint) {
+        if (leftCodePoint < 128 && rightCodePoint < 128) {
+          // ASCII comparison
+          return Integer.compare(leftCodePoint, rightCodePoint);
+        } else {
+          // substring and do UTF-8 encoded byte comparison
+          byte[] leftBytes = getUtf8SafeBytes(left, i);
```

Comment:
would it be wrong to always take `(index, index+2)`, and compare those? 

meaning, getting rid of `getUtf8SafeBytes()`

Reply:
Based on the unit test results, yes, somehow it is problematic for java, while js managed to handle it nicely.  But, increasing i one by one + `getUtf8SafeBytes()` solved the problem.